proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -3029,19 +3021,758 @@ lemma reply_unlink_ccorres:
+             apply (fastforce intro!: rf_sr_reply_update2
+                                simp: typ_heap_simps' creply_relation_def isHead_def)
+            apply ceqv
+           apply (rule ccorres_pre_threadGet, rename_tac tcbScPtrOpt)
+           apply (rule ccorres_move_c_guard_tcb)
+           apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+               apply (clarsimp simp: when_def)
+               apply (rule_tac Q="obj_at' (\<lambda>tcb. tcbSchedContext tcb = tcbScPtrOpt) tcbPtr
+                                  and valid_objs' and no_0_obj'"
+                            in ccorres_cond_both'[where Q'=\<top>])
+                 apply normalise_obj_at'
+                 apply (frule (1) obj_at_cslift_tcb)
+                 apply (frule (1) tcb_ko_at_valid_objs_valid_tcb')
+                 apply (clarsimp simp: ctcb_relation_def typ_heap_simps option_to_ptr_def
+                                       option_to_0_def valid_tcb'_def valid_bound_obj'_def)
+                 apply (rename_tac tcb tcb')
+                 apply (case_tac "tcbSchedContext tcb"; clarsimp)
+                apply (ctac (no_vcg) add: schedContext_donate_ccorres)
+               apply (rule ccorres_return_Skip)
+              apply ceqv
+             apply (rule ccorres_rhs_assoc2)
+             apply (rule ccorres_split_nothrow_novcg)
+                 apply (rule updateReply_ccorres_lemma2[where P=\<top>])
+                  apply vcg
+                 apply (frule (1) obj_at_cslift_reply)
+                 apply (fastforce intro!: rf_sr_reply_update2
+                                    simp: typ_heap_simps' creply_relation_def isHead_def
+                                   split: if_splits)
+                apply ceqv
+               apply (ctac (no_vcg) add: reply_unlink_ccorres)
+              apply (wpsimp wp: updateReply_valid_objs')
+             apply (clarsimp simp: guard_is_UNIV_def)
+            apply (rule_tac Q'="\<lambda>_ s. reply_at' replyPtr s \<and> tcb_at' tcbPtr s
+                                      \<and> valid_objs' s \<and> no_0_obj' s
+                                      \<and> pspace_aligned' s \<and> pspace_distinct' s
+                                      \<and> weak_sch_act_wf (ksSchedulerAction s) s"
+                         in hoare_post_imp)
+             apply (clarsimp simp: valid_reply'_def)
+            apply (wpsimp wp: schedContextDonate_valid_objs')
+           apply (clarsimp simp: guard_is_UNIV_def)
+          apply (rule_tac Q'="\<lambda>_ s. reply_at' replyPtr s \<and> sc_at' (theHeadScPtr (replyNext reply)) s
+                                    \<and> valid_objs' s \<and> no_0_obj' s
+                                    \<and> pspace_aligned' s \<and> pspace_distinct' s \<and> pspace_bounded' s
+                                    \<and> sym_heap_sched_pointers s \<and> valid_sched_pointers s
+                                    \<and> weak_sch_act_wf (ksSchedulerAction s) s"
```

Comment:
I'm guessing the instantiations are necessary -- would there be any potential benefit in directly assuming `invs'` at this point already? Mostly depends on what lemmas are available about the functions and how ugly they are, but sometimes it's surprisingly nice and short.

Reply:
`invs'` isn't so nice here because we'll want to start arguing about non-zombie caps and other annoying things, unfortunately.